chainstate/src/detail/tests/syncing_tests.rs

```diff
@@ -100,7 +100,7 @@ fn get_headers() {
         assert_eq!(expected[0].prev_block_id(), &Some(locator[0].get_id()));
 
         // Produce more blocks than `HEADER_LIMIT`, so get_headers is truncated.
-        btf.create_chain(&last_block.get_id(), header_limit - expected.len()).unwrap();
+        btf.create_chain(&last_block.get_id(), header_limit- expected.len()).unwrap();
```

Comment:
It's a known issue

Reply:
It's a known issue
https://github.com/AltSysrq/proptest/issues/149

I'll try to find some workaround, like this for example:
https://lib.rs/crates/proptest-attr-macro

Reply:
I spent some time digging into the `proptest-attr-macro` crate. Though I managed to make it work with ranges and `Arbitrary`, there is a blocking problem with this crate.  The configuration of test runner is hardcoded and there is no way to set custom number of test cases.

I see following options:

1. Go with !proptest macro, but change the way we arrange code in tests. Move test body to a separate function or closure, so that there is minimum code left inside the macro. This will allow rustfmt to go through the body of a test.
```
#[test]
fn test_case() {
    let test_body = |arg1: u128, arg2: u128| {
        // test body ...
    };
    proptest!(ProptestConfig::with_cases(1), |(arg1 in 1..10, agr2 in 1..20u128)| {
        test_body(arg1, arg2);
    });
} 
```
2. Fork `proptest-attr-macro` crate and add customisation. Though I'm not sure how easily it can be done.
3. Stick with proptest crate, but avoid using proptest! macro. This will require some boilerplate code to be written.

I'd probably stick with option 1, because it requires the least effort giving us full functionality of proptest.

Reply:
I'm starting to think we need option 4. Stop using proptest, and just write a function that creates a deterministic random number generator with a non-deterministic seed. Much easier than all this...

All these options seem heavily invasive and will just complicate things... all we need is reproducible random behavior.

Reply:
I also don't like the idea of getting rid of proptest entirely, though I still think that in this pull request proptest is overkill.

Reply:
I think there's a misunderstanding. I'm talking specifically about this PR because I asked Gosha to look into making random number generation deterministic. We thought proptest offers a good way for that. But then if the price of that is this massacre, then it's not worth it.

Reply:
Proptest provides a bunch of things
1. Pretty sophisticated compositional framework for test input generation
2. Automatic test case minimisation (shrinking)
3. Reproducibility of previously failed test cases by means of saving the rng seed
4. Maybe some other stuff I've forgotten

Here, we only use (3). Points (1) and (2) don't bring much benefit here. However, the `script` crate benefits from all three features a fair bit and getting rid of proptest there would be a major step back in test coverage, all because the formatter does not work properly on tests.

My preferred long-term solution would be to develop an attribute macro for proptest that has all the required features, i.e. test case count configurability and ability to specify custom generation strategies.

Regarding a more immediate term solutions, I can see the following options:
1. Attempt @azarovh's option (1), factor out test bodies. Still some (minimal) bits of code would be out of reach of the formatter.
2. Try @azarovh' option (3), go without macros. I am willing to guess that most of the boilerplate required could be abstracted away into a helper function, so the amount of extra mess in the tests themselves may be kept fairly minimal.
3. Bite the bullet for the time of being and wait with deterministic reproducibility for the proper attribute-macro based solution.

Fundamentally, the formatting issue is a problem with "scoped" (non-attribute) macros rather than proptest. The formatter is even reluctant to adjust tailing commas in `assert!`.

Reply:
> I think there's a misunderstanding. I'm talking specifically about this PR because I asked Gosha to look into making random number generation deterministic. We thought proptest offers a good way for that. But then if the price of that is this massacre, then it's not worth it.

Agreed on this random point. The question (a bit more long term) is whether we can enhance our testing by leaning into some of the other strengths proptest offers and get deterministic seeding for free.

The `script` crate does that to a degree but still suffers from the issue of formatting not working properly because everything is in a big macro. It would be nice to address that.

Reply:
We can give proptest another shot with option 3. Then we look again into how it looks and the boilerplate code price we're gonna pay. 

I'm not so excited about option 1. Channeling all the tests into multiple levels is confusing and I'd like to avoid that. Reminds me of how bitcoin does rpc... my god, what a mess! Rpc handling there is just lambdas... formatting is a mess.

Reply:
> We can give proptest another shot with option 3. Then we look again into how it looks and the boilerplate code price we're gonna pay.

You mean option 3 in @azarovh's list or in my list? Sorry for changing the numbering :laughing:

Reply:
In @azarovh's list. Since Gosha wants some practice, we can use this opportunity to improve the tests. Two birds with one stone. I don't think we're at the point of abandoning the whole thing yet. Even introducing our own deterministic seeded generator is really easy with a function... then just print the seed, and done... I already did something similar with chacha20 generator. It's not a big deal. The only question is whether we want to do it or go with proptest.

Reply:
Tried to craft a minimal working code with proptest runner but without a macro
```
#[test]
fn test_case() {
    let mut config = Config::with_cases(1);
    config.source_file = Some(file!());
    TestRunner::new(config)
        .run(&(10..20u128, 10..20u128), |(arg1, arg2)| {
            //test body
            Ok(())
        })
        .unwrap();
}
```
Imho this doesn't look convenient enough. I'm not sure that it can be generalised without macro...

At this point I think the approach with separate seedable deterministic function would be optimal solution.

Reply:
I'm OK with that, assuming @iljakuklic has no better suggestion.

Simply create a function that returns a deterministic random number generator. See in the crypto crate how the signature looks like. When creating the generator, print the seed that was used. If something goes wrong in the tests, let the dev just replace the value in that function with whatever value they found. 

One thing you can research is how to get the printing to work with logging, and not with `println!()`. This has been an annoyance to me for a while and I couldn't get to it to figure it out.

Reply:
> I'm OK with that, assuming @iljakuklic has no better suggestion.
> 
> Simply create a function that returns a deterministic random number generator. See in the crypto crate how the signature looks like. When creating the generator, print the seed that was used. If something goes wrong in the tests, let the dev just replace the value in that function with whatever value they found.
> 
> One thing you can research is how to get the printing to work with logging, and not with `println!()`. This has been an annoyance to me for a while and I couldn't get to it to figure it out.

This sounds good to me

> Tried to craft a minimal working code with proptest runner but without a macro
> 
> ```
> #[test]
> fn test_case() {
>     let mut config = Config::with_cases(1);
>     config.source_file = Some(file!());
>     TestRunner::new(config)
>         .run(&(10..20u128, 10..20u128), |(arg1, arg2)| {
>             //test body
>             Ok(())
>         })
>         .unwrap();
> }
> ```
> 
> Imho this doesn't look convenient enough. I'm not sure that it can be generalised without macro...
> 
> At this point I think the approach with separate seedable deterministic function would be optimal solution.

I think it could be trimmed down to something like:
```rust
#[test]
fn test_case() {
    run_single_proptest(file!(), &(10..20u128, 10..20u128), |(arg1, arg2)| {
        //test body
        Ok(())   // (not sure this can be eliminated, but with some trickery it probably can)
    });
}
```

Still have to specify all strategies manually and pass the result of the `file!()` macro. Not sure It's worth it.